protocols/raft/src/test/java/io/atomix/protocols/raft/RaftTest.java

```diff
@@ -1276,6 +1277,50 @@ public void testSnapshotSentOnDataLoss() throws Throwable {
     assertEquals(leaderStorage.openSnapshotStore().getCurrentSnapshotIndex(), slaveStorage.openSnapshotStore().getCurrentSnapshotIndex());
   }
 
+  @Test
+  public void testCorrectTermInSnapshot() throws Throwable {
+    final List<RaftMember> members =
+        Lists.newArrayList(createMember(), createMember(), createMember());
+    List<MemberId> memberIds = members
+            .stream()
+            .map(RaftMember::memberId)
+            .collect(Collectors.toList());
+    final Map<MemberId, RaftStorage> storages =
+        members.stream()
+            .map(RaftMember::memberId)
+            .collect(Collectors.toMap(Function.identity(), this::createStorage));
+    final Map<MemberId, RaftServer> servers =
+        storages.entrySet().stream()
+            .collect(Collectors.toMap(Map.Entry::getKey, this::createServer));
+
+    // wait for cluster to start
+    startCluster(servers);
+
+    servers.get(members.get(0).memberId())
+        .shutdown();
+
+    // fill two segments then compact so we have at least one snapshot
+    final RaftClient client = createClient(members);
+    final TestPrimitive primitive = createPrimitive(client);
+    fillSegment(primitive);
+    fillSegment(primitive);
+    MemberId leaderId = members.stream().filter(m -> servers.get(m.memberId()).isLeader()).findFirst().get().memberId();
+    servers.get(leaderId).compact().get(15_000, TimeUnit.MILLISECONDS);
+
+    final Snapshot currentSnapshot = storages.get(leaderId).openSnapshotStore().getCurrentSnapshot();
+    final long leaderTerm = servers.get(leaderId).getTerm();
+
+    assertEquals(currentSnapshot.term(), leaderTerm);
+
+    final RaftServer server = createServer(members.get(0).memberId());
+    server.join(memberIds).get(15_000, TimeUnit.MILLISECONDS);
```

Comment:
When `join` completes, we are guaranteed to have received the leader snapshot? Are we in a passive state until then because there were no events to apply?

Reply:
In this case since the sever is restarting, I assume join completes after receiving a snapshot.